programs/move_loader_api/src/lib.rs

```diff
@@ -0,0 +1,83 @@
+    224, 198, 97, 103, 136, 113, 64, 0, 0, 0, 0,
+];
+
+solana_sdk::solana_name_id!(
+    MOVE_LOADER_PROGRAM_ID,
+    "MvLdr11111111111111111111111111111111111111"
+);
+
+use language_e2e_tests::compile_and_execute;
+use log::*;
+use solana_sdk::account::KeyedAccount;
+use solana_sdk::instruction::InstructionError;
+use solana_sdk::loader_instruction::LoaderInstruction;
+use solana_sdk::pubkey::Pubkey;
+use std::str;
+use types::transaction::{Program, TransactionArgument};
+
+pub fn process_instruction(
+    _program_id: &Pubkey,
+    keyed_accounts: &mut [KeyedAccount],
+    ix_data: &[u8],
+) -> Result<(), InstructionError> {
+    solana_logger::setup();
+
+    if let Ok(instruction) = bincode::deserialize(ix_data) {
+        match instruction {
+            LoaderInstruction::Write { offset, bytes } => {
+                if keyed_accounts[0].signer_key().is_none() {
+                    warn!("key[0] did not sign the transaction");
+                    return Err(InstructionError::GenericError);
+                }
+                let offset = offset as usize;
+                let len = bytes.len();
+                debug!("Write: offset={} length={}", offset, len);
+                if keyed_accounts[0].account.data.len() < offset + len {
+                    warn!(
+                        "Write overflow: {} < {}",
+                        keyed_accounts[0].account.data.len(),
+                        offset + len
+                    );
+                    return Err(InstructionError::GenericError);
+                }
+                keyed_accounts[0].account.data[offset..offset + len].copy_from_slice(&bytes);
+            }
+            LoaderInstruction::Finalize => {
+                if keyed_accounts[0].signer_key().is_none() {
+                    warn!("key[0] did not sign the transaction");
+                    return Err(InstructionError::GenericError);
+                }
+                keyed_accounts[0].account.executable = true;
+                info!(
+                    "Finalize: account {:?}",
+                    keyed_accounts[0].signer_key().unwrap()
+                );
+            }
+            LoaderInstruction::InvokeMain { data } => {
+                if !keyed_accounts[0].account.executable {
+                    warn!("Account not executable");
+                    return Err(InstructionError::GenericError);
+                }
+
+                //// TODO: Check that keyed_accounts[0].owner is the Move program id.
+
+                // TODO: Return an error
+                let args: Vec<TransactionArgument> = bincode::deserialize(&data).unwrap();
+
+                let (programs, _params) = keyed_accounts.split_at_mut(1);
+                let program: Program = serde_json::from_slice(&programs[0].account.data).unwrap();
+                let code = str::from_utf8(&program.code()).unwrap();
+
+                compile_and_execute(&code, args).unwrap().unwrap();
```

Comment:
The prior change was closer to what we'll end up with.  In its current form address and state are assumed to be default/empty.

This should allow us to run a very simple Move program, have you tried passing the arithmetic move program in?
https://github.com/libra/libra/blob/master/language/e2e_tests/src/tests/arithmetic.rs